### ISSUE TITLE
H M L  - top / middle / bottom of page [Feature Request]

### ISSUE THREAD

**Issue by jjjchens235** [godbout/kindaVim.theapp#58]
Is there any plan to implement the keys `H` `M` `L` to navigate the current window? Thank you!

Edit: I'm hoping to use this program more extensively when editing notes in `Evernote Desktop`. However, I noticed that none of the `d` or `c` commands, i.e dw/cw work.
Is this an issue with the Evernote client, or with kV?

**Comment by godbout** [comment]
> Is there any plan to implement the keys `H` `M` `L` to navigate the current window? Thank you!

yes! but i've postponed it for a while because from my investigation it's going to be widely inconsistent. to be able to position the caret properly i need to get the visible range of characters from the Accessibility API. most apps don't seem to implement it properly, including Apple's ones. so it may get funky. will dig into it again.

> 
> Edit: I'm hoping to use this program more extensively when editing notes in `Evernote Desktop`. However, I noticed that none of the `d` or `c` commands, i.e dw/cw work. Is this an issue with the Evernote client, or with kV?

i don't use Evernote. but my guess would be that it's an Electron app then, or some custom keyboard shortcuts. you probably need to play around the `Families` Preferences and see what works best. i'll have a look at it but later today.

**Comment by godbout** [comment]
> i don't use Evernote. but my guess would be that it's an Electron app then, or some custom keyboard shortcuts. you probably need to play around the `Families` Preferences and see what works best. i'll have a look at it but later today.

so, checked and yeah. Evernote is one of those Electron apps, which means it's gonna be hit-and-miss. the Electron framework in itself doesn't support the Accessibility very well, and on top of that each app using the Electron framework can do whatever they want (and on top of that they don't use the Electron framework version). your best bet is to play with the `Families`, the last three ones.

try `Electron` first (see below). it seems some parts will work very well, while some will not. that's the problem with Electron.

![Screen Shot 2022-03-13 at 15 49 43](https://user-images.githubusercontent.com/121373/158050482-de7e512b-0428-48bc-8df2-053968a7e845.png)

if the parts you use don't work well and are important for you, maybe it's better to switch to `Key Mapping` (see below). this will use keyboard remapping rather than text manipulation. less accurate, but may work better in your case:

![Screen Shot 2022-03-13 at 15 51 11](https://user-images.githubusercontent.com/121373/158050525-53019149-51a3-4b38-b83a-bc9d8458a791.png)

last option would be 911. but usually this is for ultra emergency cases. like using a Microsoft app.

**Comment by jjjchens235** [comment]
Appreciate your detailed answer. I'm using a [legacy version ](https://help.evernote.com/hc/en-us/articles/360052560314-Install-an-older-version-of-Evernote)of Evernote that is non-Electron-based, I'll mess around with the `Key Mappings` options. Thank you again!

**Comment by godbout** [comment]
seems the legacy is some HTML content. not yet an Electron, more like basically a browser. if you put it in the `PG-R` Mode is should work pretty nicely (multipaste in a row not supported yet but coming).

**Comment by jjjchens235** [comment]
That did the trick for deleting text, thank you!!

**Comment by godbout** [comment]
unfortunately adding recalcitrant apps to the right `Families` has to be done manually. as far as macOS is concerned manipulation the text worked (in Evernote), everything went well, but it didn't. there's no reliable way to have a great UX automatically, hence having to play a bit with the `Families` when things don't go smooth. enjoy!

**Comment by godbout** [comment]
i added `H`, `L` and `M` in [kV12](https://github.com/godbout/kindaVim.theapp/releases/tag/12). (`H` and `L` with count.)

but two things:
1. when we deal with File Text  moves (i.e. that don't depend on the screen) we can be 100% accurate, but when we deal with Screen Text we depend on whatever the macOS Accessibility and/or the app is giving us back. so you'll find that those moves maybe be inaccurate sometimes, or even not work at all (like in browsers).
2. because the goal with kV is for the moves to work in all conditions (as much as the OS technically allows), all the moves are built independently, from scratch. so i've added `H`, `L`, and `M` for Normal Mode, when the Accessibility can get data. that means that Visual Mode `H`, `L`, `M` are not included. neither are `cH`, `dL`, `yM` etc. those moves have to be built separately. if you need them let me know and i'll add them to the priority list. thanks!

**Comment by jjjchens235** [comment]
Awesome, when trying to update the app, I'm getting this error:
<img width="259" alt="image" src="https://user-images.githubusercontent.com/19519336/158811826-a6e1dac6-2ac4-489f-9961-c2d20cac8723.png">

**Comment by godbout** [comment]
you're probably running it from the `Download` folder. you need to move it to your `Applications` folder. in any case, you could just also remove the old version and download the new. all your settings are stored on your computer, so they'll still be there.

**Comment by jjjchens235** [comment]
Shoot I had written a rather long message typed out already but must have failed to press enter yest, let me try again.

First of all, thank you for implementing this feature, it's been working wonderfully.

A few points:
1. The only time this implementation seems to fail is if I'm on the very last line `G`
2. I was hoping to use this set of commands after scrolling down via `<C-D>`. Let's say I scrolled down 50x from the top, and then pressed 'M', I would hope to be still scrolled down 50x, but just in the middle of the screen. However, it takes me back to the top of the page (but in the middle now). This is not an issue with 'M', but with the scroll-down functionality in general.

To get point 2 fixed, should I open a separate issue? Lastly, is there anything on the roadmap to implement `zz zb zt` ?

**Comment by godbout** [comment]
> Shoot I had written a rather long message typed out already but must have failed to press enter yest, let me try again.

the tribulations of using computers 
> 
> First of all, thank you for implementing this feature, it's been working wonderfully.
> 
> A few points:
> 
> 1. The only time this implementation seems to fail is if I'm on the very last line `G`
> 2. I was hoping to use this set of commands after scrolling down via `<C-D>`. Let's say I scrolled down 50x from the top, and then pressed 'M', I would hope to be still scrolled down 50x, but just in the middle of the screen. However, it takes me back to the top of the page (but in the middle now). This is not an issue with 'M', but with the scroll-down functionality in general.
> 
> To get point 2 fixed, should I open a separate issue? Lastly, is there anything on the roadmap to implement `zz zb zt` ?

**Comment by godbout** [comment]
i've created a new issue for that here: https://github.com/godbout/kindaVim.theapp/issues/60
comment there or "follow"(?) to get updates.

**Comment by jjjchens235** [comment]
Here is the full text:
```

Wang Lang Market ตลาดวังหลัง - Wang Lang Market is one my favorite snacking and browsing markets in Bangkok. 
Chatuchak Weekend Market ตลาดนัดจตุจักร - One of the most popular things to do in Bangkok is go shopping - don't miss the Chatuchak Weekend Market.
Klong Toey Market ตลาดคลองเตย - The most vital fresh food market in Bangkok.
Floating Market ตลาดน้ำ - In this video I visit Talad Nam Bang Nam Pheung floating market in southern Bangkok.
Pratunam Market ประตูน้ำ - Shopping is a huge Bangkok attraction and Pratunam is one of the wholesale outdoor markets. 
MBK / Siam Shopping มาบุญครอง - One of the most hectic shopping malls is MBK, everything imaginable under one roof. 
Or Tor Kor Market ตลาด อ.ต.ก. - Visiting markets is one of my personal favorites activities in Bangkok, and Or Tor Kor is amazing. 
Dusit Zoo สวนสัตว์ดุสิต - The Dusit Zoo is one of the most fun things to do in Bangkok if you are with children.
Lumpini Park สวนลุมพินี- The central park of Bangkok.
Massage นวดแผนไทย- Don't miss a rejuvenating massage in Thailand!
Grand Palace / Wat Phra Kaew วัดพระเก้า- The most revered of all attractions in Bangkok.
Wat Pho วัดโพธิ์ - Home of the reclining Buddha and the birthplace of the traditional Thai massage.
Wat Arun วัดอรุณ - Wat Arun is the temple of dawn, located on the the Chao Phraya River. 
at Saket (วัดสระเกศ) - Bangkok's Golden Mountain Temple with great views.
Erawan Museum ช้างสามเศียร- A giant three headed elephant which is a temple and museum.
Vimanmek Mansion พระที่นั่งวิมานเมฆ - As the largest golden teak wood mansion in the world, it's very impressive.
Khao San Road ถนนข้าวสาร - There are lots of things to do in this area, many of them revolving around parties and nightlife.
Silom and Patpong (สีลม) - Silom at night turns into a market with Patpong street, home to many go-go bars and fake goods.
Victory Monument อนุสาวรีย์ชัยสมรภูมิ - One of the transportation hubs of Bangkok and there's a lot of food, including the tasty boat noodle alley.
Eat Durian กินทุเรียน - Durian, also known as the king of fruits, is creamy and sweet, and it's the best fruit in the world.
Thai Street Food อาหารข้างทาง - Eating Thai street food is one of the most rewarding things to do in Bangkok, mainly because there's so much to try, and so many good flavors.
Thai Cooking Class เรียนทำอาหาร- A cooking class is a great way to learn some Thai recipes!
Pahurat พาหุรัด- Bangkok's little India is known as Pahurat. It's a place to shop for fabric and dine on delicious Indian food.
Pak Klong Talad ปากคลองตลาด- The largest flower market in Bangkok is a great place to see colorful and exotic flowers all day long.
Yaowarat / Sampeng Market (เยาวราช / สำเพ็ง) - Yaowarat is one of the busiest and most exciting places in Bangkok - don't miss it!


Victory Monument อนุสาวรีย์ชัยสมรภูมิ - One of the transportation hubs of Bangkok and there's a lot of food, including the tasty boat noodle alley.
Eat Durian กินทุเรียน - Durian, also known as the king of fruits, is creamy and sweet, and it's the best fruit in the world.
Thai Street Food อาหารข้างทาง - Eating Thai street food is one of the most rewarding things to do in Bangkok, mainly because there's so much to try, and so many good flavors.
Thai Cooking Class เรียนทำอาหาร- A cooking class is a great way to learn some Thai recipes!
Pahurat พาหุรัด- Bangkok's little India is known as Pahurat. It's a place to shop for fabric and dine on delicious Indian food.
Pak Klong Talad ปากคลองตลาด- The largest flower market in Bangkok is a great place to see colorful and exotic flowers all day long.
Yaowarat / Sampeng Market (เยาวราช / สำเพ็ง) - Yaowarat is one of the busiest and most exciting places in Bangkok - don't miss it

```

Besides the original issue of the last line not working, `M` and `L` don't necessarily go where I want it go go.

For example, here's a screenshot of my full screen, i would expect `M`  to land me around the red area, but I end up on the line `Patpong` (which is below the screenshot).

<img width="503" alt="image" src="https://user-images.githubusercontent.com/19519336/159118257-1d5d2787-3a77-4175-a50e-47b114b42197.png">

**Comment by godbout** [comment]
ha i love Bangkok, and Thailand in general. would spend a few months every year there before Covid. when i still had a passport. Thai food is my fav.

> Besides the original issue of the last line not working, `M` and `L` don't necessarily go where I want it go go.
> 
> For example, here's a screenshot of my full screen, i would expect `M` to land me around the red area, but I end up on the line `Patpong` (which is below the screenshot).
> 
> <img alt="image" width="503" src="https://user-images.githubusercontent.com/19519336/159118257-1d5d2787-3a77-4175-a50e-47b114b42197.png">

so is this happening in Evernote Legacy Edition? it's important for me to know because as i've mentioned earlier, those moves require apps and macOS to send correct Screen data. when we deal with moves like `G` and you do a `25G`, as long as i have access to the text, i can control the calculations myself (counting linefeeds, whitespaces etc). but when we deal with moves like `H` `L` `M` `CTRL-D` `zz` etc., we need to rely on what the app and macOS tell us about what part of the text is visible on the screen. i'll check but that's probably the issue here. i need to know which app it is though.

the visible range of character is a pain and i don't think that this part of the AX API works properly. for example you can see this if you use `G`/`gg` in long text in browsers. even if i can properly calculate the caret position and set the caret there, the text doesn't scroll. all those stuff have been reported to Apple (e.g.: https://github.com/godbout/AXEngine/blob/b2c1cac18658e098322d061c66f991335e13c215/Sources/AXEngine/AXEngine.swift#L299) but it's a 20+ years old old clunky API that hasn't changed since then. so for those Screen moves, as said, it's unfortunately a hit-and-miss, and the technology available is what's gonna limit the accuracy or feasibility here.

**Comment by jjjchens235** [comment]
Going to Bangkok/Chiang Mai in June and very excited for the Thai food :)

Correct, Evernote legacy

**Comment by godbout** [comment]
> Going to Bangkok/Chiang Mai in June and very excited for the Thai food :)

![angry-frustrated-black-kid-congrats-happy-for-you-nice](https://user-images.githubusercontent.com/121373/159211675-335bc516-34bf-40c5-9ce4-cf9aa92bc646.jpg)

😂😂😂

> Correct, Evernote legacy

yeah, Evernote Legacy is sending back wrong data. this is not a TextArea, this is a bunch of Static Texts put together to look like a TextArea. 

<img width="1280" alt="Screen Shot 2022-03-21 at 14 00 59" src="https://user-images.githubusercontent.com/121373/159211982-f01d204a-5421-40a7-917a-eaecac2397f3.png">

getting the whole text and acting on it is ok. but when it's about Screen data, those "apps" return wrong data. nothing i can do.

regarding the last line, it works in other apps. so it's a Evernote thing again. although it doesn't work if the last line is empty. if you remove that empty line, it will work.

if you try `H` `L` `M` in proper apps like Notes or iA Writer you'll see that the moves work. although `L` and `M` "scroll" the text also, which should not happen. this is something related to the `CTRL-D` etc. moves, that i will investigate. it's about that visible character range attribute that is clunky. unfortunately you're asking for the moves that are limited not due to my knowledge or skills but because of macOS API's limitations. i'll try to push forward but it's possible that this is not doable with consistency.

**Comment by godbout** [comment]
i've improve `H` `L` and `M` so that they stop scrolling. but it's still not going to work in Evernote Legacy. at whatever location you are in that app, it will send you back the start of the text. so those 3 moves will always end up in the same place, whatever the current scrolling. i guess the way to get this better would be to ask Evernote to implement the Accessibility properly.

**Comment by godbout** [comment]
confirmed. see below:

https://user-images.githubusercontent.com/121373/159241235-11dc3f67-fbef-4470-b3d3-fd528a9e5086.mp4

Evernote Legacy only returns the range of the first view. it never updates. so we can't do anything proper with this. only solution is for Evernote to fix this. i doubt they will as it's the Legacy edition. new version uses Electron, which is pretty bad too in general. some apps make the effort to handle the Accessibility quite ok, but most don't. and quite ok is still pretty bad.

**Comment by jjjchens235** [comment]
HAHA. I actually read on your website you've been stuck in Macau, super crazy.

I confirmed that HML works great in `TextEdit`. As you said, it's only the scrolling issue, which maybe is outside of scope.

 It's too bad this doesn't work in Evernote, I guess have no choice but to write them a letter.

I appreciate you taking such a detailed look into this issue.

**Comment by godbout** [comment]
> HAHA. I actually read on your website you've been stuck in Macau, super crazy.

even 10k kilometers away the French gov can still fuck with you.

> I confirmed that HML works great in `TextEdit`. As you said, it's only the scrolling issue, which maybe is outside of scope.
> 
> It's too bad this doesn't work in Evernote, I guess have no choice but to write them a letter.

the scrolling issue will be "fixed" in the next release. in the sense that it will stop scrolling. but that part of macOS is still not very good. the visible range returned by the AX API is usually not that accurate, and the caret ends up out of visibility. then when you use `j`/`k` to move the caret the input gets scrolled automatically so that the caret gets in the middle. i'll do what is possible to be doing with the APIs but unfortunately those are gonna be the first moves that are not going to be accurate ☹️

all the `CTRL-D` and similar have to use the same API. so it's not gonna work that properly too. i'll build them but not sure if it's gonna be useful to you. for example i have `CTRL-D` moving the caret half a page down, but i can't scroll so that the caret appears on the top, even when setting the visible range (i've sent feedback to Apple a year ago about this; will do another one). this will affect also `zt` and `zb`. `zz` should look ok as it's centered by nature.

> I appreciate you taking such a detailed look into this issue.

hey, thank YOU! those were in the plans but i knew they would be a bit funky, so it was not a priority. thanks to you they're getting done, which ultimately makes kindaVim better. so, thanks again.

**Comment by jjjchens235** [comment]
I wish I could say something moderately insightful but you lost me lol.

If there's no good solution, there's no need to implement right away (at least for me), I'm in no huge rush, I already appreciate `H M L` :)

**Comment by godbout** [comment]
:D

it's just that Vim feels straightforward when you're used to it, but when i started building each move from scratch, then i saw that there's way more going on than we realize. so to get a move in macOS working like in Vim, you need to understand first exactly how it works. and if it doesn't behave exactly like in Vim, it's going feel very off. even more with those moves that play with scrolling.

anyways. having `H` `L` `M` updated. and adding `CTRL-D U B F`. plus some other stuff. should be out in a couple of days.

**Comment by godbout** [comment]
13 is out: https://github.com/godbout/kindaVim.theapp/releases/tag/13
it stops scrolling for `H` `L` `M`. if it works fine we can close this issue i think. the `CTRL-D` etc. and scrolling for them will be reviewed in their relevant issues.

**Comment by godbout** [comment]
ok after using them, i want to change `H` `L` `M` :D it's regarding what i've mentioned before. i need your input. sorry but you have to do a bit of work :D

i'm gonna try to explain more clearly. video coming.

**Comment by godbout** [comment]
the first video is the current implementation. **it is how Vim works**. it makes the calculation for `H L M` according to file (text) lines, that means lines delimited by linefeed, however they appear on a screen. 

https://user-images.githubusercontent.com/121373/160269316-99bf2cef-f242-42ed-bf12-858ab6946ab0.mp4

you can see that it behaves as you would expect as long as the lines are not too long, but once you have file (text) lines that span over multiple screen lines, the positioning seems to be off. this is because Vim sets the position according to file (text) lines. the difference between Vim and kindaVim is that Vim controls its own buffer, so it always show a text starting with a file (text) line. kindaVim/macOS doesn't. so it feels off.

**Comment by godbout** [comment]
now what i want to do is this:

https://user-images.githubusercontent.com/121373/160269472-219ec40b-86c8-4018-bda9-e8e13517a0c1.mp4

not caring about file (text) lines, but always putting `H L M` in the top, bottom or middle of the screen, however the length of the file (text) lines. **THIS IS NOT VIM COMPLIANT**. Vim does not behave like this. but it feels better to me. but i'm also not a pure Vim hardcore user, i'm more of a Vim Mode from other text editors user. hence the need for input.

**Comment by jjjchens235** [comment]
I'm in agreement with you, the 2nd option is more user-friendly even though it doesn't mimic Vim's internal behavior as closely.

**Comment by jjjchens235** [comment]
Thinking about this more though, the way `H L M` are implemented currently (like VIM) is fine.

The issue is the orientation is not adjusted when scrolling up/down. Would your new solution fix this?

**Comment by godbout** [comment]
> The issue is the orientation is not adjusted when scrolling up/down. Would your new solution fix this?

sorry not sure what you mean. can you elaborate?

**Comment by jjjchens235** [comment]
I had written this 5 days ago:

>  I confirmed that HML works great in TextEdit. As you said, it's only the scrolling issue, which maybe is outside of scope.

I have no issue with how `HML` is currently implemented per the above. The main issue was that `HML` wasn't working as expected after scrolling with `<C-D>`.

However, looking at this more, this was an issue with `Evernote` specifically, it works as expected in TextEdit. This is `kindavim12` btw.

-----

Ok, I have also tried `kindavim13` now.  `HML` and `<C-D>` both work well still. I think the difference is now `<C-D` is smoother in that I'm now centered each scroll which is pretty cool!

In summary, I think the issues I pointed out originally were Evernote specific.

**Comment by godbout** [comment]
> I have no issue with how `HML` is currently implemented per the above. The main issue was that `HML` wasn't working as expected after scrolling with `<C-D>`.

ok. the issue was actually `<C-D>`, that was not implemented until kV13 for the Accessibility Strategy. more below.

> Ok, I have also tried `kindavim13` now. `HML` and `<C-D>` both work well still. I think the difference is now `<C-D` is smoother in that I'm now centered each scroll which is pretty cool!

so yes, `<C-D>` for the Accessibility Strategy was made in kV13: https://github.com/godbout/kindaVim.theapp/releases/tag/13
Accessibility Strategy means text manipulation, so starting from kV13, when you're in an input that handles the macOS Accessibility, then the text can be read and i can accurately calculate on which character should the caret ends. before that, in kV12 and below, it would just remap `<C-D>` to `PageDown`. it will still do this in anything that is not text.

so `<C-D>` in text is way better now yes. but the scrolling is still not similar to Vim. in Vim the text would scroll but visually your caret wouldn't move. it's like if the text would be scrolling *behind* the caret. doing this will require more work and it's gonna be the same work needed for `zz zt zb`, although for those 3 ones the caret moves with the scrolling but stays on the same character, like if it was stuck *on* the text. opposite goals, but same technology.

tell me if i talk too much coz i tend to explain everything and dump my head like A LOT 😅

> In summary, I think the issues I pointed out originally were Evernote specific.

yes. kV is best in apps that send proper Accessibility information (caret position, selected text, selected length, visible area, etc). then it can do ok with apps that don't send Accessibility information by remapping keys. the apps it does not well are apps that send **wrong** Accessibility information. usually Electron based. that's why there's the `Families` settings in the Preferences. Evernote is one of those bad ones.

**Comment by godbout** [comment]
> Edit: One small point. In `textedit`, `H` gets me 1-3 lines higher than I would expect (even on single lines), and `L` gets me one-3 lines lower than I would expect. Not a big deal for me, but calling it out in case that's unexpected behavior.

yeah i saw that for `L`. actually the last video i showed, that issue is already fixed 😅 it will be available in kV14.

for `H`, sorry i have to repeat myself, but that's gonna be the issue with moves that are not text based but screen based. with text i can calculate accurately where to put the caret, with screen i have to trust what macOS gives me. the issue you see with TextEdit is probably because you have the ruler option set. whether you have it on or off, macOS doesn't seem to care. when asked what is the text area that is visible, it sends the same info with both options. of course if you have the ruler on, that means that caret will be behind. i'm having the same issue with Xcode and all the tabs at the top. see the video below (seems buying ScreenFlow was a good investment 😂):

https://user-images.githubusercontent.com/121373/160431893-3103506a-d3ab-40eb-9531-6b67bff879a2.mp4

**Comment by jjjchens235** [comment]
Ah ok, ruler setting in textedit makes sense. 
>> ultimately not sure whether i'll keep H M L working in text lines or if i'll change to screen lines. screen lines, even though it's not Vim compliant, that feels way better. digest and see.

Sounds good to me!

**Comment by godbout** [comment]
to get proper scrolling for all the moves you're asking about the engines need quite an upgrade. it's another part of macOS i need to learn and experiment with but if i can get it to work consistently that's gonna bring a plethora of new moves. thanks for pushing the boundaries and making kV better! 💪

**Comment by godbout** [comment]
ok `H` and `L` are a bit better now, in kV14. should most of the time be on the top or bottom lines that are fully visible. still counting on File Lines, not Screen Lines though. will see later if i change. next is gonna be `zz zb zt` and the alike. if all is good here we can close this issue? i'll open one for `zz` etc.

**Comment by jjjchens235** [comment]
Awesome, kv14 feels great, thank you!

**Comment by jjjchens235** [comment]
`zz` etc seem to be working perfect! But `H` seems to be working a little weird now.

`H` goes to the expected line at the top, BUT it centers the text which is unexpected. `M` `L` have no such issue.

**Comment by godbout** [comment]
i can reproduce yeah. the tests didn't catch that because it works fine if the sentences are short enough. but not good when it's long lines yeah.

https://user-images.githubusercontent.com/121373/162555811-7194ec91-3cc3-4a58-bb2f-eac2b1b18a5b.mp4

i'll check and hopefully fix for next release. was it working fine before?

**Comment by godbout** [comment]
ok, added a test so that the issue doesn't come back: https://github.com/godbout/AccessibilityStrategyTestApp/commit/99f28f2bd608dbb525f5f40412e371088d6eb5de

been improved. will be released in kV16, hopefully within a week. thanks for the report!

**Comment by godbout** [comment]
kV16 out 🔥️ if i didn't fuck up it should be all fixed in there. let me know! if yes then we can close here.

**Comment by godbout** [comment]
https://github.com/godbout/kindaVim.theapp/releases/tag/16

**Comment by jjjchens235** [comment]
works good :)
